### ISSUE TITLE
Cannot create stacked bar chart without aggregate

### ISSUE THREAD

**Issue by rueckstiess** [vega/vega-lite#2595]
I cannot seem to create a stacked chart with data that I have already pre-aggregated outside of vega-lite. I don't know if this is by design or a bug. All examples of stacked bar/area charts I can find use some kind of aggregate in the `y` channel.

But the docs don't mention this. From the docs on [Stacked Bar Chart](https://vega.github.io/vega-lite/docs/mark.html#stacked-bar-chart): 
> Adding color to the bar chart (by using the color attribute) creates a stacked bar chart by default." 

Here is a simple spec with data that I would expect to create a stacked bar chart: 
```json
{
  "width": 300,
  "height": 300,
  "data": {
    "values": [
      {
        "dayOfWeek": 0,
        "payment_type": "Credit Card",
        "tips": 53
      },
      {
        "dayOfWeek": 0,
        "payment_type": "Cash",
        "tips": 93
      },
      {
        "dayOfWeek": 0,
        "payment_type": "No Charge",
        "tips": 4
      },
      {
        "dayOfWeek": 1,
        "payment_type": "Credit Card",
        "tips": 48
      },
      {
        "dayOfWeek": 1,
        "payment_type": "Cash",
        "tips": 71
      },
      {
        "dayOfWeek": 1,
        "payment_type": "No Charge",
        "tips": 1
      },
      {
        "dayOfWeek": 2,
        "payment_type": "Credit Card",
        "tips": 49
      },
      {
        "dayOfWeek": 2,
        "payment_type": "Cash",
        "tips": 77
      },
      {
        "dayOfWeek": 2,
        "payment_type": "No Charge",
        "tips": 1
      }
    ]
  },
  "mark": "bar",
  "encoding": {
    "x": {
      "field": "dayOfWeek",
      "type": "nominal"
    },
    "y": {
      "field": "tips",
      "type": "quantitative"
    },
    "color": {
      "field": "payment_type",
      "type": "nominal"
    }
  }
}
```

I'm getting this chart:

![screen shot 2017-06-28 at 17 16 14](https://user-images.githubusercontent.com/99221/27624803-9d21dd12-5c25-11e7-837f-477fa4ee3d0a.png)

However if I add a `aggregate: "sum"` to the `y` channel (which is a no-op as all the combinations of `payment_type` and `dayOfWeek` are unique), it shows the expected result: 

![screen shot 2017-06-28 at 17 18 12](https://user-images.githubusercontent.com/99221/27624867-ddb3569e-5c25-11e7-9de7-df5a01608ef7.png)

This seems to be an easy enough workaround, but I'd like to check if this is by design, in which case it would be great to document this behavior, e.g. "for stacked charts, you always need an aggregate".

**Comment by domoritz** [comment]
Yes, we usually don't expect pre-aggeragted data. For now, please just add a placeholder aggregate.

Another related issue is allowing pre-binned data in Vega-Lite.

**Comment by rueckstiess** [comment]
Thanks for the quick answer. I can use that work-around for now, and manually overwrite the axis label to get rid of the `SUM(...)`. FYI, this is for MongoDB Charts, and we are planning to do all aggregations directly on the database server (via the aggregation framework) for performance reasons and just pass the results into vega-lite, so that's why it is pre-aggregated. :-)

**Comment by kanitw** [comment]
> Yes, we usually don't expect pre-aggregated data.

I don't think that was the reason it was implemented this way. 
I think the original reason was that the stack operator still "sums" the tips displayed in the stack.
That said, I see that even sum is not presented, automatically stacking (which inherently applies sum) is still a better option.

**Comment by pelotom** [comment]
Our company also preaggregates the data in a database, I think this might be quite a common pattern in industrial-scale situations.

**Comment by domoritz** [comment]
Preaggregated and prebinned data will be my focus after the 2.0 release. I
agree that it's super important for scalability.

**Comment by rueckstiess** [comment]
Great, thanks @domoritz! Once you get to preaggregation/prebinning, I'd be happy to have a chat about what we do at MongoDB with vega and how this could be generalized for other databases. 

Currently we parse the vega-lite script, extract all aggregation/binning operations that we can handle in the database, pre-aggregate the data, and then update the spec to make vega-lite display it correctly. This is a bit hacky (e.g. required to use a "sum" aggregate when it's already summed up) and may not always work.

Ideally there would be a way to declare which operations can be handled by a database server, and which operations should be done on the client in vega/vega-lite.

**Comment by kanitw** [comment]
> This is a bit hacky (e.g. required to use a "sum" aggregate when it's already summed up) and may not always work.

There are two possible cases for stack:

1) If you have already sum the value and apply stack operator in the database, then this case is already supported.  For example, this [spec](https://github.com/vega/vega-lite/blob/master/examples/specs/bar_diverging_stacked.vl.json) just uses bar with `x` and `x2` -- this should be obvious and not hacky. 

2) If you have already the sum the value of the field and have not apply stack operator and want Vega-Lite to do the stack calculation, then in a way, the whole bars are still showing the sum of the sums that you have already calculated.  So semantically it is not a hack too. 

Personally, I do not think that database have to implement a stack operator since if the data that Vega-Lite have to sum has already been summed, it should be cheap enough to do stack calculation on the client side. 

> Currently we parse the vega-lite script, extract all aggregation/binning operations that we can handle in the database, pre-aggregate the data, and then update the spec to make vega-lite display it correctly. 

Good point.  I have been thinking about this for a bit too, but let's separate this conversation to another issue https://github.com/vega/vega-lite/issues/3069.

**Comment by kanitw** [comment]
>  and may not always work.

Could you please explain a case where it does not work?  Maybe I have overlooked something.

**Comment by pelotom** [comment]
@kanitw to me it’s not a question of whether the hack works, it’s about the obvious and concise syntax not doing what you expect. There is no clear reason to me why stacking behavior should be limited to aggregated fields.

**Comment by jheer** [comment]
I agree that this is highly unintuitive. In addition, it appears to depend on the specific aggregate function applied. For example, in the online editor this spec (which changes `sum` to `min`) removes the stacking:

```json
{
  "$schema": "https://vega.github.io/schema/vega-lite/v2.json",
  "data": { "url": "data/population.json"},
  "transform": [
    {"filter": "datum.year == 2000"},
    {"calculate": "[null,'Male','Female'][datum.sex]", "as": "gender"}
  ],
  "mark": "bar",
  "encoding": {
    "y": {
      "aggregate": "min", "field": "people", "type": "quantitative",
      "axis": {"title": "population", "grid": false}
    },
    "x": {
      "field": "age", "type": "ordinal"
    },
    "color": {
      "field": "gender", "type": "nominal",
      "scale": {"range": ["#EA98D2", "#659CCA"]}
    }
  },
  "config": {
    "view": {"stroke": "transparent"}
  }
}
```

I would advocate for a bit less "magic" here in favor of more consistent defaults.

**Comment by kanitw** [comment]
> There is no clear reason to me why stacking behavior should be limited to aggregated fields.

I agree with this part.  

> For example, in the online editor this spec (which changes sum to min) removes the stacking.

I think we can just throw warning instead

**Comment by kanitw** [comment]
Fixed in https://github.com/vega/vega-lite/pull/3070

**Comment by hanzigs** [comment]
May I know whether this is possible now, i.e stacked bar chart without aggregate?

**Comment by kanitw** [comment]
Yes, as discussed above this is fixed in #3070. If you look into #3070, you would see that we add `stacked_bar_unaggregate` example in the PR.

**Comment by hanzigs** [comment]
@kanitw Thank you for the reply
Can I get help on this https://stackoverflow.com/questions/73411763/vega-lite-infinite-extent-for-field-field-start-infinity-infinity
